File: 20-Mediator/src/go/chat_room.go
package main

// ChatRoom is the Concrete Mediator — it holds the colleagues and owns the
// coordination logic: broadcast to everyone except the sender.
type ChatRoom struct {
	users []User
}

func (c *ChatRoom) AddUser(user User) {
	c.users = append(c.users, user)
}

func (c *ChatRoom) SendMessage(message string, sender User) {
	for _, user := range c.users {
		if user != sender {
			user.Receive(message, sender.Name())
		}
	}
}

File: 20-Mediator/src/rust/chat_room.rs
use std::rc::Rc;

use crate::user::User;

/// Concrete Mediator — holds the colleagues and owns the coordination logic:
/// broadcast each message to everyone except the sender.
pub struct ChatRoom {
    users: Vec<Rc<User>>,
}

impl ChatRoom {
    pub fn new() -> Self {
        ChatRoom { users: Vec::new() }
    }

    pub fn add_user(&mut self, user: Rc<User>) {
        self.users.push(user);
    }

    pub fn send_message(&self, message: &str, sender_name: &str) {
        for user in &self.users {
            if user.name() != sender_name {
                user.receive(message, sender_name);
            }
        }
    }
}
